Hw8

tiginamaria opened the pull request.

tiginamaria committed:
project init

tiginamaria committed:
added implementation for all methods TODO shutdown

tiginamaria committed:
added tests, fixed thread bugs

tiginamaria committed:
added tests for exceptions and singleton

tiginamaria committed:
changed the structure of Task in to make faster thenApply execution

yurii-litvinov commented on ThreadPool/.idea/uiDesigner.xml:
```diff
@@ -0,0 +1,124 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Это лучше не выкладывать

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/LightExecutionException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.hse.hw;
+
+public class LightExecutionException extends Exception {
```
Надо комментарии и для исключения тоже

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/LightFuture.java:
```diff
@@ -0,0 +1,29 @@
+     * @param <R> type of the new task's result
+     * @return new task
+     */
+    <R> LightFuture<R> thenApply(Function<T, R> function);
```
Можно более мягкие ограничения на тип функции

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+     * Count current active working threads
+     */
+    public int countActiveThreads() {
+        return  Thread.activeCount();
```
Я бы не сказал, что этот метод имеет какое-то отношение к ThreadPool, так что его тут не стоит реализовывать

tiginamaria replied:
Он нужен для тестов.

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+        public void run() {
+            Task<?> task;
+            while (!Thread.interrupted()) {
+                synchronized(taskQueue) {
```
Пробела не хватает

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+     * @return LightFuture - box, where result of task is stored
+     */
+    public <T> LightFuture<T> execute(@NotNull Supplier<T> supplier) {
+        Task<T> task = new Task<>(supplier);
```
var?

yurii-litvinov replied:
var :)

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+        for (var thread : threadPool) {
+            thread.interrupt();
+        }
+    }
```
Если потом кто-то сделает execute, ему вернут LightFuture, который никогда не закончит вычисление просто потому, что все потоки остановлены и делать его некому. Вызов get у такого LightFuture приведёт к дедлоку

yurii-litvinov replied:
Кроме того, хочется, чтобы shutdown возвращал управление только когда все потоки уже полностью остановлены.

yurii-litvinov replied:
Тут лучше сначала всем разослать interrupt(), затем всем сделать join(). Иначе какая-нить вычислительно сложная задача может сильно замедлить процесс.

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+         */
+        @Override
+        public T get() throws LightExecutionException, InterruptedException {
+            while(!isReady.get()) {
```
Тоже пробела не хватает

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,238 @@
+            synchronized (thenApplyTasks) {
+                synchronized (taskQueue) {
+                    taskQueue.addAll(thenApplyTasks);
+                    taskQueue.notifyAll();
```
Как-то это делается в обход ThreadPoolImpl-а, это может привести к проблемам с shutdown и вообще стрёмно :)

yurii-litvinov commented on ThreadPool/src/test/java/ru/hse/hw/ThreadPoolImplTest.java:
```diff
@@ -0,0 +1,145 @@
+        int size = new Random().nextInt(1000);
+        ArrayList<Integer> array = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            array.add(new Random().nextInt());
```
Этак у Вас все значения в массиве могут оказаться одинаковыми. На некоторых платформах (где нет /dev/random) конструктор может использовать в качестве сида значение системного таймера в миллисекундах, а этот цикл легко уместится в одну миллисекунду даже несмотря на то, что это Java.

Кроме того, сборщик мусора тысяче Random-ов не обрадуется. Стандартный паттерн --- использовать один Random на весь класс или даже на всё приложение.

yurii-litvinov commented on ThreadPool/src/test/java/ru/hse/hw/ThreadPoolImplTest.java:
```diff
@@ -0,0 +1,145 @@
+
+    private ArrayList<Integer> getRandomIntegerArray() {
+        int size = new Random().nextInt(1000);
+        ArrayList<Integer> array = new ArrayList<>(size);
```
var?

tiginamaria committed:
thread pool mistakes fixed
added join to shutdown and flag to con continue execution after shut down
added comments to LightExecutionException

tiginamaria committed:
fixed code stile mistakes

tiginamaria committed:
thread putting thenApply directly to queue mistakes fixed

tiginamaria committed:
deleted usless files

tiginamaria committed:
deleted countActiveThreads

tiginamaria committed:
travis added

tiginamaria committed:
Merge branch 'master' into hw8

tiginamaria committed:
make travis work

yurii-litvinov commented on Phonebook/pom.xml:
```diff
@@ -5,7 +5,7 @@
     <modelVersion>4.0.0</modelVersion>
```
Что-то лишние файлы в диффе

yurii-litvinov commented on ThreadPool/src/main/java/ru/hse/hw/ThreadPoolImpl.java:
```diff
@@ -0,0 +1,259 @@
+                setException(e);
+            }
+
+            isReady.set(true);
```
Тут set делается не под замком, поэтому могут быть гонки с thenApply --- там мы рискуем добавить задачи в очередь как раз в тот момент, когда тут сделается isReady.set(true) и задачи никто не исполнит (synchronized (thenApplyTasks) в executeThenApplyLightFuture тоже не очень спасёт, решение ведь принимается на основе thenApply).
